### DISCUSSION TITLE
How to make directional block in minecraft 1.18.2

### DISCUSSION THREAD

**Discussion by Dog517** [FabricMC/fabric-api#2159] [Mod Dev Support] [answered]
Hello, im trying to make a directional block for minecraft. here is my register block code :

>     public static final Block OBAMIUM_BLOCK = registerBlock("obamium_block", new Block(FabricBlockSettings.of(Material.METAL).strength(10.0f).requiresTool()), ItemGroup.MATERIALS);

here is my blockstate.json:

> {
  "variants": {
    "facing=north": { "model": "stuffmod:block/obamium_block"},
    "facing=east":  { "model": "stuffmod:block/obamium_block", "y":  90},
    "facing=south": { "model": "stuffmod:block/obamium_block", "y": 180},
    "facing=west":  { "model": "stuffmod:block/obamium_block", "y": 270}
  }
}

here is my block model.json:

> {
  "parent": "block/cube",
  "textures": {
    "down": "stuffmod:block/bottom",
    "up": "stuffmod:block/back_or_top",
    "north": "stuffmod:block/front",
    "east": "stuffmod:block/sideleft",
    "south": "stuffmod:block/back_or_top",
    "west": "stuffmod:block/sideright",
    "particle": "stuffmod:block/bottom"
  }
}

and here is the item model.json:

> {
  "parent": "stuffmod:block/obamium_block"
}

Thanks for any help!

**Comment by warjort** [comment]
You need to set the properties for your block in its constructor.
The properties are what you specify as the keys of the variants in your blockstates.

See for example the PillarBlock used for wood logs which uses the AXIS property.
But I guess you are looking to use the net.minecraft.state.property.Properties.FACING property.

**Reply by Dog517** [reply]
where can i find the pillar block for a example?

**Reply by warjort** [reply]
Like I said, look at Minecraft's PillarBlock or a block that references that FACING property.
If you don't have the Minecraft source in your ide run gensources for your gradle project.

There is also this, but it might not be up-to-date for 1.18.2?
https://fabricmc.net/wiki/tutorial:directionalblock

**Reply by Dog517** [reply]
I made the pillarblock class and gave it to my block like so:

> package net.dogdev.stuffmod.block.custom;
import net.minecraft.block.AbstractBlock;
import net.minecraft.block.Block;
import net.minecraft.block.BlockState;
import net.minecraft.item.ItemPlacementContext;
import net.minecraft.state.StateManager;
import net.minecraft.state.property.EnumProperty;
import net.minecraft.state.property.Properties;
import net.minecraft.state.property.Property;
import net.minecraft.util.BlockRotation;
import net.minecraft.util.math.Direction;
import net.minecraft.util.math.Direction.Axis;
public class ModObamaBlock extends Block {
    public static final EnumProperty<Direction.Axis> AXIS;

    public ModObamaBlock(AbstractBlock.Settings settings) {
        super(settings);
        this.setDefaultState((BlockState)this.getDefaultState().with(AXIS, Axis.Y));
    }

    public BlockState rotate(BlockState state, BlockRotation rotation) {
        return changeRotation(state, rotation);
    }

    public static BlockState changeRotation(BlockState state, BlockRotation rotation) {
        switch (rotation) {
            case COUNTERCLOCKWISE_90:
            case CLOCKWISE_90:
                switch ((Direction.Axis)state.get(AXIS)) {
                    case X:
                        return (BlockState)state.with(AXIS, Axis.Z);
                    case Z:
                        return (BlockState)state.with(AXIS, Axis.X);
                    default:
                        return state;
                }
            default:
                return state;
        }
    }

    protected void appendProperties(StateManager.Builder<Block, BlockState> builder) {
        builder.add(new Property[]{AXIS});
    }

    public BlockState getPlacementState(ItemPlacementContext ctx) {
        return (BlockState)this.getDefaultState().with(AXIS, ctx.getSide().getAxis());
    }

    static {
        AXIS = Properties.AXIS;
    }
}

and gave it to my block:

> public static final Block OBAMIUM_BLOCK = registerBlock("obamium_block", new ModObamaBlock(FabricBlockSettings.of(Material.METAL).strength(10.0f).requiresTool()), ItemGroup.MATERIALS);

but its still a black and purple cube.

>

**Reply by warjort** [reply]
The AXIS property is not the FACING property.

They have different properties. Look at its enum values.
Axis is x,y,z
Facing is north,south,etc.
Either change your blockstates to use the axis values as keys or change your block to use the FACING property.

Also look at your logs/latest.log it will have some warning saying why it was unable to load your model. The black/purple just means that there was a problem.

**Reply by Dog517** [reply]
could you post the ModObamaBlock Class modified to be the FACING property? Im not smart enough :)

**Reply by warjort** [reply]
Just scan replace AXIS -> FACING

Then choose a default facing in the constructor, e.g. north.
````
this.setDefaultState(this.stateManager.getDefaultState().with(FACING, Direction.NORTH))
````

For the getPlacementState() look at a block that uses the FACING property, e.g. the ChestBlock which simplifies to something like:
````
Direction direction = ctx.getPlayerFacing().getOpposite();
return this.getDefaultState().with(FACING, direction);
````

If you don't understand this basic stuff, you won't have a clue with harder things. :-)

**Reply by Dog517** [reply]
so, I scan replaced AXIS with FACING, but im still getting errors from this bit of code:

>                 switch ((Direction.Facing)state.get(FACING)) {
                    case X:
                        return (BlockState)state.with(FACING, FACING.Z);
                    case Z:
                        return (BlockState)state.with(FACING, FACING.X);
                    default:
                        return state;
X and Ys are red, and so is Facing in Direction.Facing

**Reply by warjort** [reply]
that code is for axis rotation you copied it from the pillar block
its not relevant for the facing property, remove the rotate/changeRotation methods

**Reply by Dog517** [reply]
Cannot resolve method 'getFacing()' at 

>  public BlockState getPlacementState(ItemPlacementContext ctx) {
        return (BlockState)this.getDefaultState().with(FACING, ctx.getSide().getFacing());
    }

**Reply by warjort** [reply]
see my previous comment for the implementation of that method

my simple suggestion was make the block face the player
````
Direction direction = ctx.getPlayerFacing().getOpposite();
return this.getDefaultState().with(FACING, direction);
````
But, it is up to you how you really want to implement it.

**Reply by Dog517** [reply]
here is the github repository for the mod. If you could make a fork or something that would be good.
https://github.com/Dog517/fabric-StuffMod-mod-1.18/tree/master/src/main

**Reply by warjort** [reply]
I am not going to implement your mod for you.

I can answer questions when you get stuck, 
but you need to take the time to try to understand what is going on.

If you need full code example(s), look at the wiki, the vanilla minecraft blocks or other mods with similar blocks on github.

**Reply by Dog517** [reply]
ok ill try to figure it out.

**Comment by warjort** [comment]
You also need to add() your properties to the StateManager in the block's appendProperties().
Again, see the PillarBlock as a simple example.

**Comment by Dog517** [comment, marked as answer]
I fixed it! I just copied the AbstractFurnaceItem to my custom block classs!
